<!DOCTYPE html>
<html lang="zh-CN">
<head><!-- hexo injector head_begin start -->
<link rel="stylesheet" href="/css/bilicard.css">
<!-- hexo injector head_begin end -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <meta name="keywords" content="空中有零，零域，分享快乐，音乐，心得，生活，电影，代码，编程，考试，学业，零式的天空，凌空博客，原零空博客，原Tans博客，现零的小屋官方博客，计算机，音乐，电脑，编程，生活，分享">
    <meta name="description" content="一个技术控，爱技术，爱分享，爱音乐">
    <meta name="author" content="Pin Young">
    <script defer data-pjax src="/js/iosurl.js"></script>
    
    <title>
        
            基因注释 |
        
        零域
    </title>
    
<link rel="stylesheet" href="/css/style.css">

    <link rel="shortcut icon" href="/img/favicon.png">
    
<link rel="stylesheet" href="/font/css/fontawesome.min.css">

    
<link rel="stylesheet" href="/font/css/regular.min.css">

    
<link rel="stylesheet" href="/font/css/solid.min.css">

    
<link rel="stylesheet" href="/font/css/brands.min.css">

    
    <script id="hexo-configurations">
    let KEEP = window.KEEP || {}
    KEEP.hexo_config = {"hostname":"0skyu.cn","root":"/","language":"zh-CN","path":"search.xml"}
    KEEP.theme_config = {"toc":{"enable":true,"number":true,"expand_all":true,"init_open":true},"style":{"primary_color":"#EE5A24","logo":"/img/favicon.png","favicon":"/img/favicon.png","avatar":"img/1.jpg","font_size":null,"font_family":null,"hover":{"shadow":true,"scale":false},"first_screen":{"enable":true,"header_transparent":false,"background_img":"/img/bg/333.jpg","description":"Keep writing and Keep loving.","font_color":"#f0d21a","hitokoto":true},"scroll":{"progress_bar":true,"percent":false}},"local_search":{"enable":true,"preload":true},"code_copy":{},"code_block":{"tools":{"enable":true,"style":"default"},"highlight_theme":"default"},"side_tools":{},"pjax":{"enable":true},"lazyload":{"enable":true},"comment":{"enable":false,"use":"gitalk","valine":{"appid":null,"appkey":null,"server_urls":null,"placeholder":null},"gitalk":{"github_id":"DsTansice","github_admins":"DsTansice","repository":"Null-Comment","client_id":"b109b534b2268088d780","client_secret":"7808e138a1ffbcd9f02bb39cc1ee7fb41828c52f","proxy":null},"twikoo":{"env_id":"twikoo-0skyucn-1gkzaji32bb03146","region":null,"version":"1.6.8"},"waline":{"server_url":null,"reaction":false,"version":2}},"post":{"author_label":{"enable":true,"auto":true,"custom_label_list":["Trainee","Engineer","Architect"]},"word_count":{"enable":true,"wordcount":true,"min2read":true},"img_align":"left","copyright_info":true},"version":"3.6.1"}
    KEEP.language_ago = {"second":"%s 秒前","minute":"%s 分钟前","hour":"%s 小时前","day":"%s 天前","week":"%s 周前","month":"%s 个月前","year":"%s 年前"}
    KEEP.language_code_block = {"copy":"复制代码","copied":"已复制","fold":"折叠代码块","folded":"已折叠"}
    KEEP.language_copy_copyright = {"copy":"复制版权信息","copied":"已复制","title":"原文标题","author":"原文作者","link":"原文链接"}
  </script>
<meta name="generator" content="Hexo 6.3.0"><link rel="alternate" href="/atom.xml" title="零域" type="application/atom+xml">
</head>


<body>
<div class="progress-bar-container">
    
        <span class="scroll-progress-bar"></span>
    

    
        <span class="pjax-progress-bar"></span>
        <i class="pjax-progress-icon fas fa-circle-notch fa-spin"></i>
    
</div>


<main class="page-container">

    

    <div class="page-main-content">

        <div class="page-main-content-top">
            
<header class="header-wrapper">

    <div class="header-content">
        <div class="left">
            
                <a class="logo-image" href="/">
                    <img src="/img/favicon.png">
                </a>
            
            <a class="logo-title" href="/">
               零域
            </a>
        </div>

        <div class="right">
            <div class="pc">
                <ul class="menu-list">
                    
                        <li class="menu-item">
                            <a class=""
                               href="/"
                            >
                                首页
                            </a>
                        </li>
                    
                        <li class="menu-item">
                            <a class=""
                               href="/archives"
                            >
                                归档
                            </a>
                        </li>
                    
                        <li class="menu-item">
                            <a class=""
                               href="/tags"
                            >
                                标签
                            </a>
                        </li>
                    
                        <li class="menu-item">
                            <a class=""
                               href="/categories"
                            >
                                分类
                            </a>
                        </li>
                    
                        <li class="menu-item">
                            <a class=""
                               href="/links"
                            >
                                友链
                            </a>
                        </li>
                    
                        <li class="menu-item">
                            <a class=""
                               target="_blank" rel="noopener" href="https://hao.0skyu.cn/"
                            >
                                HAO
                            </a>
                        </li>
                    
                    
                        <li class="menu-item search search-popup-trigger">
                            <i class="fas fa-search"></i>
                        </li>
                    
                </ul>
            </div>
            <div class="mobile">
                
                    <div class="icon-item search search-popup-trigger"><i class="fas fa-search"></i></div>
                
                <div class="icon-item menu-bar">
                    <div class="menu-bar-middle"></div>
                </div>
            </div>
        </div>
    </div>

    <div class="header-drawer">
        <ul class="drawer-menu-list">
            
                <li class="drawer-menu-item flex-center">
                    <a class=""
                       href="/">首页</a>
                </li>
            
                <li class="drawer-menu-item flex-center">
                    <a class=""
                       href="/archives">归档</a>
                </li>
            
                <li class="drawer-menu-item flex-center">
                    <a class=""
                       href="/tags">标签</a>
                </li>
            
                <li class="drawer-menu-item flex-center">
                    <a class=""
                       href="/categories">分类</a>
                </li>
            
                <li class="drawer-menu-item flex-center">
                    <a class=""
                       href="/links">友链</a>
                </li>
            
                <li class="drawer-menu-item flex-center">
                    <a class=""
                       target="_blank" rel="noopener" href="https://hao.0skyu.cn/">HAO</a>
                </li>
            
        </ul>
    </div>

    <div class="window-mask"></div>

</header>


        </div>

        <div class="page-main-content-middle">

            <div class="main-content">

                
                    <div class="fade-in-down-animation">
    <div class="post-page-container">
        <div class="article-content-container">

            <div class="article-title">
                <span class="title-hover-animation">基因注释</span>
            </div>

            
                <div class="article-header">
                    <div class="avatar">
                        <img src="/img/1.jpg">
                    </div>
                    <div class="info">
                        <div class="author">
                            <span class="name">Pin Young</span>
                            
                                <span class="author-label">Lv9</span>
                            
                        </div>
                        <div class="meta-info">
                            
<div class="article-meta-info">
    <span class="article-date article-meta-item">
        
            <i class="fa-regular fa-calendar-plus"></i>&nbsp;
        
        <span class="pc">2020-04-30 20:49:05</span>
        <span class="mobile">2020-04-30 20:49</span>
    </span>
    
        <span class="article-update-date article-meta-item">
        <i class="fas fa-file-pen"></i>&nbsp;
        <span class="pc">2023-02-05 21:25:09</span>
    </span>
    
    
    
        <span class="article-tags article-meta-item">
            <i class="fas fa-tags"></i>&nbsp;
            <ul>
                
                    <li>
                        <a href="/tags/%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A/">基因注释</a>&nbsp;
                    </li>
                
            </ul>
        </span>
    

    
    
        <span class="article-wordcount article-meta-item">
            <i class="fas fa-file-word"></i>&nbsp;<span>1.9k 字</span>
        </span>
    
    
        <span class="article-min2read article-meta-item">
            <i class="fas fa-clock"></i>&nbsp;<span>7 分钟</span>
        </span>
    
    
        <span class="article-pv article-meta-item">
            <i class="fas fa-eye"></i>&nbsp;<span id="busuanzi_value_page_pv"></span>
        </span>
    
</div>

                        </div>
                    </div>
                </div>
            

            <div class="article-content keep-markdown-body">
                

                <h1 id="基因注释"><a href="#基因注释" class="headerlink" title="基因注释"></a>基因注释</h1><p>记录下自己对RNA-seq基因注释的学习,并对Drop-seq软件包中的注释模块进行代码研读</p>
<h2 id="什么是基因注释"><a href="#什么是基因注释" class="headerlink" title="什么是基因注释"></a>什么是基因注释</h2><p>一句话概况注释:<em>找到与reads有overlap的基因片段,并进行标记</em></p>
<p>这里reads指bam文件中的每一行数据,即测序下机文件fastq与参考基因组进行比对之后生成的数据,其中记录了每条read在参考基因组中的位置,有起始位置和终止位置,表示一段区间</p>
<p>基因注释文件记录了每个基因片段在参考基因组上的位置,也是一段区间,因此与bam文件结合,通过find overlapping我们可以查找到每条read属于哪个基因片段,将其标记在bam格式的tags中,这对后续的生信分析是有帮助的</p>
<h2 id="基因注释文件"><a href="#基因注释文件" class="headerlink" title="基因注释文件"></a>基因注释文件</h2><p>GTF&#x2F;GFF格式是基因注释的常用格式<br>GTF是Gene Transfer Format的缩写,其文件由九列数据组成,以tab分割,示例如下:</p>
<table>
<thead>
<tr>
<th>seq_id</th>
<th>source</th>
<th>type</th>
<th>start</th>
<th>end</th>
<th>score</th>
<th>strand</th>
<th>phase</th>
<th>attributes</th>
</tr>
</thead>
<tbody><tr>
<td>chr1</td>
<td>HAVANA</td>
<td>exon</td>
<td>11869</td>
<td>12227</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>gene_id “ENSG00000223972.5”; transcript_id “ENST00000456328.2”;…</td>
</tr>
<tr>
<td>染色体编号</td>
<td>注释来源</td>
<td>类型</td>
<td>在参考序列的起始位置</td>
<td>终止位置</td>
<td>得分,说明注释信息的可能性</td>
<td>位于参考序列的正负链</td>
<td>仅对类型为CDS有效,表示起始编码的位置</td>
<td>包含众多属性的列表</td>
</tr>
</tbody></table>
<p>虽然数据有九列之多,但并不是所有都会用到,常用的有:</p>
<ol>
<li>seq_id. 要查找overlap,首先得是同一条染色体</li>
<li>type. 有多种类型,如gene&#x2F;transcript&#x2F;exon&#x2F;CDS&#x2F;UTR等,它们之间有层级关系,一般gtf文件中多行数据对应一条基因的完整信息,以type为gene的行为起始;每条gene可以表示为树状结构,gene为根节点,第二层为transcript,第三层为诸如exon CDS等</li>
<li>start&#x2F;end. 根据起始终止位置可以建立interval,这是find overlapping的基础</li>
<li>strand. 正负链可以作为过滤条件,假如一条read与多个基因有overlap,可以根据方向是否相同过滤掉部分基因</li>
<li>attributes. 一些列键值对属性,常用的信息包括名称,id之类</li>
</ol>
<h2 id="注释流程分析"><a href="#注释流程分析" class="headerlink" title="注释流程分析"></a>注释流程分析</h2><p>流程可分为三步:</p>
<ol>
<li><strong>读入gtf文件</strong>. 从磁盘将gtf文件加载进内存,并提取需要的信息,毕竟gtf有许多信息是我们不需要的</li>
<li><strong>建立区间树</strong>. 即interval tree,使用区间树是为了高效查询,为了达到最佳性能,一般使用基于红黑树的区间树实现,因为红黑树是平衡树,查找时间复杂度O(lgN),不会出现退化成链表的最坏情况</li>
<li><strong>查找区间并注释</strong>. 遍历bam文件中每条read,根据其在参考序列中的位置构建interval,与前面建立的interval tree进行overlap的查找,找到之后,进行一些逻辑计算,并更新read的tags,输出到bam</li>
</ol>
<h2 id="Drop-seq代码研读"><a href="#Drop-seq代码研读" class="headerlink" title="Drop-seq代码研读"></a>Drop-seq代码研读</h2><p>Drop-seq是使用java开发的程序包,其中的<em>TagReadWithGeneExonFunction</em> 模块实现了添加注释的功能</p>
<h3 id="主流程"><a href="#主流程" class="headerlink" title="主流程"></a>主流程</h3><p>主流程核心代码为:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 加载基因注释文件,并构建区间树</span></span><br><span class="line"><span class="keyword">final</span> OverlapDetector&lt;Gene&gt; geneOverlapDetector = GeneAnnotationReader.loadAnnotationsFile(ANNOTATIONS_FILE, bamDict);</span><br><span class="line"><span class="comment">// 打开输出bam文件</span></span><br><span class="line">SAMFileWriter writer= <span class="keyword">new</span> <span class="title class_">SAMFileWriterFactory</span>().makeSAMOrBAMWriter(header, <span class="literal">true</span>, OUTPUT);</span><br><span class="line"><span class="comment">// 遍历输入bam中的每条read</span></span><br><span class="line"><span class="keyword">for</span> (SAMRecord r: inputSam) &#123;</span><br><span class="line">    <span class="comment">// 对完成比对的read,进行find overlapping操作并添加注释</span></span><br><span class="line">    <span class="keyword">if</span> (!r.getReadUnmappedFlag())</span><br><span class="line">        r=setAnnotations(r, geneOverlapDetector);</span><br><span class="line">    <span class="comment">// 输出注释后的read</span></span><br><span class="line">    writer.addAlignment(r);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>其结果是根据overlap的genes信息,添加三个Tag,示例:</p>
<blockquote>
<p>GE:Z:WASH7P    XF:Z:CODING        GS:Z:-</p>
</blockquote>
<ul>
<li>GE为gene name  </li>
<li>XF为locus function  </li>
<li>GS为正负链</li>
</ul>
<h3 id="加载gtf并构建interval-tree"><a href="#加载gtf并构建interval-tree" class="headerlink" title="加载gtf并构建interval tree"></a>加载gtf并构建interval tree</h3><p>核心代码:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 用来解析gtf每行数据,提取需要的字段</span></span><br><span class="line"><span class="keyword">final</span> <span class="type">FilteringGTFParser</span> <span class="variable">parser</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">FilteringGTFParser</span>(gtfFlatFile);</span><br><span class="line"><span class="comment">// gene name相同的gtf行,代表它们是一个gene的数据,使用它们构建GeneFromGTF数据类型,</span></span><br><span class="line"><span class="comment">// 此类型继承自Interval类型</span></span><br><span class="line"><span class="keyword">final</span> <span class="type">GeneFromGTFBuilder</span> <span class="variable">geneBuilder</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">GeneFromGTFBuilder</span>(parser);</span><br><span class="line"><span class="comment">// 初始化interval tree</span></span><br><span class="line"><span class="keyword">final</span> OverlapDetector&lt;GeneFromGTF&gt; overlapDetector = <span class="keyword">new</span> <span class="title class_">OverlapDetector</span>&lt;&gt;(<span class="number">0</span>, <span class="number">0</span>);</span><br><span class="line"><span class="keyword">while</span> (geneBuilder.hasNext())&#123;</span><br><span class="line">        <span class="comment">// 将每条gene添加到interval tree,其内部按照chromosome进行分类</span></span><br><span class="line">        <span class="type">GeneFromGTF</span> <span class="variable">gene</span> <span class="operator">=</span> geneBuilder.next();</span><br><span class="line">        overlapDetector.addLhs(gene, gene);</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>

<p>代码实现使用了迭代器,减少内存的消耗</p>
<p>将gtf每行数据以gene_name为key,放入<code>map&lt;gene_name, List&lt;GTFRecord&gt;&gt;</code>中,这样就将每条gene的数据分类好了<br>geneBuilder 是个<code>iter&lt;List&lt;GTFRecord&gt;&gt;</code>,迭代时,对每个gene将其数据<code>List&lt;GTFRecord&gt;</code> 按gene_version分类成<code>map&lt;gene_version, List&lt;GTFRecord&gt;&gt;</code>,对key也就是所有的gene_version进行排序,取最大的gene_version对应的<code>List&lt;GTFRecord&gt;</code>来构建GeneFromGTF(对部分gtf文件它们的gene_version是null,则对gene_version分类和没做一样)</p>
<p>重点是构建GeneFromGTF类实例:<br><code>List&lt;GTFRecord&gt;</code>转GeneFromGTF调用接口<code>makeGeneFromMultiVersionGTFRecords()</code></p>
<ol>
<li>使用list中第一条GTFRecord的信息初始化GeneFromGTF(因为第一条的类型永远是gene),只有start end属性是取得list中所有数据的最小start,最大end  </li>
<li>进行一致性检查. 检查list中所有数据,如正反链必须都一致,chr一致等,否则抛出异常</li>
<li>将所有的非gene数据进行统计处理,更新GeneFromGTF成员变量<code>Map&lt;String, TranscriptFromGTF&gt; transcripts</code>, TranscriptFromGTF类型包括transcript, CDS, coding等相关信息</li>
</ol>
<p>最后调用<em>htsjdk</em>库的<code>OverlapDetector.addLhs()</code>将GeneFromGTF作为节点加入线段树中</p>
<h3 id="注释逻辑"><a href="#注释逻辑" class="headerlink" title="注释逻辑"></a>注释逻辑</h3><p>核心代码:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> SAMRecord <span class="title function_">setAnnotations</span> <span class="params">(<span class="keyword">final</span> SAMRecord r, <span class="keyword">final</span> OverlapDetector&lt;Gene&gt; geneOverlapDetector)</span> &#123;</span><br><span class="line">        Map&lt;Gene, LocusFunction&gt; map = AnnotationUtils.getInstance().getLocusFunctionForReadByGene(r, geneOverlapDetector);</span><br><span class="line">        Set&lt;Gene&gt; exonsForRead = AnnotationUtils.getInstance().getConsistentExons (r, map.keySet(), ALLOW_MULTI_GENE_READS);</span><br><span class="line"></span><br><span class="line">        List&lt;Gene&gt; genes = <span class="keyword">new</span> <span class="title class_">ArrayList</span>&lt;&gt;();</span><br><span class="line"></span><br><span class="line">        <span class="keyword">for</span> (Gene g: exonsForRead) &#123;</span><br><span class="line">            <span class="type">LocusFunction</span> <span class="variable">f</span> <span class="operator">=</span> map.get(g);</span><br><span class="line">            <span class="keyword">if</span> (f==LocusFunction.CODING || f==LocusFunction.UTR)</span><br><span class="line">                genes.add(g);</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        List&lt;LocusFunction&gt; allPassingFunctions = <span class="keyword">new</span> <span class="title class_">ArrayList</span>&lt;&gt;();</span><br><span class="line">        <span class="keyword">if</span> (USE_STRAND_INFO) &#123;</span><br><span class="line">            <span class="comment">// constrain gene exons to read strand.</span></span><br><span class="line">            genes = getGenesConsistentWithReadStrand(genes, r);</span><br><span class="line">            <span class="comment">// only retain functional map entries that are on the correct strand.</span></span><br><span class="line">            <span class="keyword">for</span> (Gene g: map.keySet()) &#123;</span><br><span class="line">                <span class="type">boolean</span> strandCheck=readAnnotationMatchStrand(g, r);</span><br><span class="line">                <span class="keyword">if</span> (strandCheck) allPassingFunctions.add(map.get(g));</span><br><span class="line">            &#125;</span><br><span class="line">        &#125; <span class="keyword">else</span></span><br><span class="line">            allPassingFunctions=<span class="keyword">new</span> <span class="title class_">ArrayList</span>&lt;&gt;(map.values());</span><br><span class="line"></span><br><span class="line">        <span class="comment">// if strand tag is used, only add locus function values for passing genes.</span></span><br><span class="line">        <span class="keyword">for</span> (Gene g: genes)</span><br><span class="line">            allPassingFunctions.add(map.get(g));</span><br><span class="line"></span><br><span class="line">        <span class="type">LocusFunction</span> <span class="variable">f</span> <span class="operator">=</span> AnnotationUtils.getInstance().getLocusFunction(allPassingFunctions, <span class="literal">false</span>);</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> (genes.size()&gt;<span class="number">1</span> &amp;&amp; <span class="built_in">this</span>.ALLOW_MULTI_GENE_READS==<span class="literal">false</span>)</span><br><span class="line">            log.error(<span class="string">&quot;There should only be 1 gene assigned to a read for DGE purposes.&quot;</span>);</span><br><span class="line"></span><br><span class="line">        <span class="type">String</span> <span class="variable">finalGeneName</span> <span class="operator">=</span> getCompoundGeneName(genes);</span><br><span class="line">        <span class="type">String</span> <span class="variable">finalGeneStrand</span> <span class="operator">=</span> getCompoundStrand(genes);</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> (f!=<span class="literal">null</span>)</span><br><span class="line">            r.setAttribute(<span class="built_in">this</span>.FUNCTION_TAG, f.toString());</span><br><span class="line">        <span class="keyword">if</span> (finalGeneName!=<span class="literal">null</span> &amp;&amp; finalGeneStrand!=<span class="literal">null</span>) &#123;</span><br><span class="line">            r.setAttribute(<span class="built_in">this</span>.TAG, finalGeneName);</span><br><span class="line">            r.setAttribute(<span class="built_in">this</span>.STRAND_TAG, finalGeneStrand);</span><br><span class="line">        &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">            r.setAttribute(<span class="built_in">this</span>.TAG, <span class="literal">null</span>);</span><br><span class="line">            r.setAttribute(<span class="built_in">this</span>.STRAND_TAG, <span class="literal">null</span>);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">return</span> (r);</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>

<p>概况一下注释逻辑:对read构建interval,查找overlap的所有基因,计算三个tag: locusFunction, geneName, geneStrand,并更新read  </p>
<p>对每个gene计算locusFunction的核心代码:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">private</span> LocusFunction <span class="title function_">getLocusFunctionForRead</span> <span class="params">(<span class="keyword">final</span> SAMRecord rec, <span class="keyword">final</span> Gene g)</span> &#123;</span><br><span class="line">    List&lt;AlignmentBlock&gt; alignmentBlocks = rec.getAlignmentBlocks();</span><br><span class="line"></span><br><span class="line">    LocusFunction [] blockSummaryFunction = <span class="keyword">new</span> <span class="title class_">LocusFunction</span>[alignmentBlocks.size()];</span><br><span class="line">    Set&lt;Gene&gt; temp = <span class="keyword">new</span> <span class="title class_">HashSet</span>&lt;&gt;();</span><br><span class="line">    temp.add(g);</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> (<span class="type">int</span> i=<span class="number">0</span>; i&lt;alignmentBlocks.size(); i++) &#123;</span><br><span class="line">        <span class="type">AlignmentBlock</span> <span class="variable">alignmentBlock</span> <span class="operator">=</span>alignmentBlocks.get(i);</span><br><span class="line"></span><br><span class="line">        LocusFunction [] blockFunctions=getLocusFunctionsByBlock(alignmentBlock, temp);</span><br><span class="line">        <span class="type">LocusFunction</span> <span class="variable">blockFunction</span> <span class="operator">=</span> getLocusFunction(blockFunctions, <span class="literal">false</span>);</span><br><span class="line">        blockSummaryFunction[i]=blockFunction;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="type">LocusFunction</span> <span class="variable">readFunction</span> <span class="operator">=</span> getLocusFunction(blockSummaryFunction, <span class="literal">false</span>);</span><br><span class="line">    <span class="keyword">return</span> readFunction;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>计算过程:</p>
<ol>
<li>对read每个位点都计算一个LocusFunction(计算过程具体为是否与exons有overlap)</li>
<li>将一条alignmentBlocks中所有位点的LocusFunctions合并为一个LocusFunction</li>
<li>将所有alignmentBlocks的LocusFunctions合并为一个LocusFunction</li>
</ol>
<p>每次合并都是取最大的LocusFunction,其是一个枚举变量,由小到大为:<br><code>INTERGENIC, INTRONIC, UTR, CODING, RIBOSOMAL</code></p>
<h3 id="输出统计信息"><a href="#输出统计信息" class="headerlink" title="输出统计信息"></a>输出统计信息</h3><table>
<thead>
<tr>
<th>统计项</th>
<th>解释</th>
</tr>
</thead>
<tbody><tr>
<td>AMBIGUOUS_READS_REJECTED</td>
<td>有多条同方向的overlaped gene</td>
</tr>
<tr>
<td>READ_AMBIGUOUS_GENE_FIXED</td>
<td>overlaped gene中同向的只有一条,但是还有反向的</td>
</tr>
<tr>
<td>READS_RIGHT_STRAND</td>
<td>只有一条overlaped gene,并且与read同向</td>
</tr>
<tr>
<td>READS_WRONG_STRAND</td>
<td>overlaped gene没有同向,只有反向</td>
</tr>
<tr>
<td>TOTAL_READS</td>
<td>处理的总reads数</td>
</tr>
</tbody></table>
<h2 id="参考"><a href="#参考" class="headerlink" title="参考"></a>参考</h2><ul>
<li><a class="link"   target="_blank" rel="noopener" href="https://blog.csdn.net/sinat_38163598/article/details/72851239" >GTF基因注释文件详解<i class="fas fa-external-link-alt"></i></a></li>
<li><a class="link"   target="_blank" rel="noopener" href="https://www.jianshu.com/p/0800a07cfa37" >液滴微流控获取单细胞及Drop-seq_tools分析流程<i class="fas fa-external-link-alt"></i></a></li>
<li><a class="link"   target="_blank" rel="noopener" href="https://github.com/broadinstitute/Drop-seq" >Drop-seq_github<i class="fas fa-external-link-alt"></i></a></li>
</ul>

            </div>

            
                <div class="post-copyright-info">
                    
<div class="article-copyright-info-container">
    <ul class="copyright-info-content">
        <li class="post-title">
            <span class="type">本文标题</span>：<span class="content">基因注释</span>
        </li>
        <li class="post-author">
            <span class="type">本文作者</span>：<span class="content">Pin Young</span>
        </li>
        <li class="post-time">
            <span class="type">创建时间</span>：<span class="content">2020-04-30 20:49:05</span>
        </li>
        <li class="post-link">
            <span class="type">本文链接</span>：<span class="content">posts/8d7a.html</span>
        </li>
        <li class="post-license">
            <span class="type">版权声明</span>：<span class="content">本博客所有文章除特别声明外，均采用 <a class="license" target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh">BY-NC-SA</a> 许可协议。转载请注明出处！</span>
        </li>
    </ul>
    <div class="copy-copyright-info flex-center tooltip" data-content="复制版权信息" data-offset-y="-2px">
        <i class="fa-solid fa-copy"></i>
    </div>
</div>

                </div>
            

            
                <ul class="post-tags-box">
                    
                        <li class="tag-item">
                            <a href="/tags/%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A/">#基因注释</a>&nbsp;
                        </li>
                    
                </ul>
            

            
                <div class="article-nav">
                    
                        <div class="article-prev">
                            <a class="prev"
                               rel="prev"
                               href="/posts/4432.html"
                            >
                            <span class="left arrow-icon flex-center">
                              <i class="fas fa-chevron-left"></i>
                            </span>
                                <span class="title flex-center">
                                <span class="post-nav-title-item">考研复试英语口语模板推荐</span>
                                <span class="post-nav-item">上一篇</span>
                            </span>
                            </a>
                        </div>
                    
                    
                        <div class="article-next">
                            <a class="next"
                               rel="next"
                               href="/posts/2574.html"
                            >
                            <span class="title flex-center">
                                <span class="post-nav-title-item">堆的基本操作（C 语言版）</span>
                                <span class="post-nav-item">下一篇</span>
                            </span>
                                <span class="right arrow-icon flex-center">
                              <i class="fas fa-chevron-right"></i>
                            </span>
                            </a>
                        </div>
                    
                </div>
            

            
        </div>

        
            <div class="toc-content-container">
                <div class="post-toc-wrap">
    <div class="post-toc">
        <ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A"><span class="nav-number">1.</span> <span class="nav-text">基因注释</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E4%BB%80%E4%B9%88%E6%98%AF%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A"><span class="nav-number">1.1.</span> <span class="nav-text">什么是基因注释</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6"><span class="nav-number">1.2.</span> <span class="nav-text">基因注释文件</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%B3%A8%E9%87%8A%E6%B5%81%E7%A8%8B%E5%88%86%E6%9E%90"><span class="nav-number">1.3.</span> <span class="nav-text">注释流程分析</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Drop-seq%E4%BB%A3%E7%A0%81%E7%A0%94%E8%AF%BB"><span class="nav-number">1.4.</span> <span class="nav-text">Drop-seq代码研读</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%B8%BB%E6%B5%81%E7%A8%8B"><span class="nav-number">1.4.1.</span> <span class="nav-text">主流程</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%8A%A0%E8%BD%BDgtf%E5%B9%B6%E6%9E%84%E5%BB%BAinterval-tree"><span class="nav-number">1.4.2.</span> <span class="nav-text">加载gtf并构建interval tree</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E6%B3%A8%E9%87%8A%E9%80%BB%E8%BE%91"><span class="nav-number">1.4.3.</span> <span class="nav-text">注释逻辑</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E8%BE%93%E5%87%BA%E7%BB%9F%E8%AE%A1%E4%BF%A1%E6%81%AF"><span class="nav-number">1.4.4.</span> <span class="nav-text">输出统计信息</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%8F%82%E8%80%83"><span class="nav-number">1.5.</span> <span class="nav-text">参考</span></a></li></ol></li></ol>
    </div>
</div>

            </div>
        
    </div>
</div>


                
            </div>

        </div>

        <div class="page-main-content-bottom">
            
<footer class="footer">
    <div class="info-container">
        <div class="copyright-info info-item">
            &copy;
            
                <span>2014</span> -
            
            2023
            
                &nbsp;<i class="fas fa-heart icon-animate"></i>
                &nbsp;<a href="/">Pin Young</a>
            
        </div>
        
            <script async data-pjax
                    src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
            <div class="website-count info-item">
                
                    访问人数&nbsp;<span id="busuanzi_value_site_uv"></span>&ensp;
                
                
                    总访问量&nbsp;<span id="busuanzi_value_site_pv"></span>
                
            </div>
        
        <div class="theme-info info-item">
            由 <a target="_blank" href="https://hexo.io">Hexo</a> 驱动&nbsp;|&nbsp;主题&nbsp;<a class="theme-version" target="_blank" href="https://github.com/dstansice">零域Lite 1.3.1</a>
        </div>
        
        
        <script async defer data-website-id="b3b1feba-ef0f-4367-ba34-6a0152b8749b" src="https://analysis.0skyu.cn/umami.js"></script>
         
            <div class="icp-info info-item">
            <div class="theme-info info-item">
                <a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=12011002021027" style="display:inline-block;text-decoration:none;height:20px;line-height:20px;"><img src="https://0skyu.cn/beian.png" style="float:left;"/><p style="float:left;height:20px;line-height:20px;margin: 0px 0px 0px 5px; color:#939393;">津公网安备 12011002021027号</p></a> 
                </div>
                 <div class="theme-info info-item">
               <center> <a target="_blank" rel="nofollow"
                   href="https://beian.miit.gov.cn"
                >
                    津ICP备2021004482号-1
                </a></center>
                </div>
                
            </div>
        
        
            <div class="deploy-info info-item">
                
                    <a target="_blank" rel="nofollow" href="https://hao.0skyu.cn/">
                
                    本站由 <span class="tooltip" data-content="腾讯云"><img src="/images/deploy-provider/tencent_cloud.png"></span> 提供部署服务
                
                    </a>
                
            </div>
        
    </div>
</footer>

        </div>
    </div>

    
        <div class="post-tools">
            <div class="post-tools-container">
    <ul class="tools-list">
        <!-- TOC aside toggle -->
        
            <li class="tools-item flex-center toggle-show-toc">
                <i class="fas fa-list"></i>
            </li>
        

        <!-- go comment -->
        
    </ul>
</div>

        </div>
    

    <div class="right-bottom-side-tools">
        <div class="side-tools-container">
    <ul class="side-tools-list">
        <li class="tools-item tool-font-adjust-plus flex-center">
            <i class="fas fa-search-plus"></i>
        </li>

        <li class="tools-item tool-font-adjust-minus flex-center">
            <i class="fas fa-search-minus"></i>
        </li>

        <li class="tools-item tool-dark-light-toggle flex-center">
            <i class="fas fa-moon"></i>
        </li>

        <!-- rss -->
        
            <li class="tools-item rss flex-center">
                <a class="flex-center"
                   href="/atom.xml"
                   target="_blank"
                >
                    <i class="fas fa-rss"></i>
                </a>
            </li>
        

        
            <li class="tools-item tool-scroll-to-top flex-center">
                <i class="fas fa-arrow-up"></i>
            </li>
        

        <li class="tools-item tool-scroll-to-bottom flex-center">
            <i class="fas fa-arrow-down"></i>
        </li>
    </ul>

    <ul class="exposed-tools-list">
        <li class="tools-item tool-toggle-show flex-center">
            <i class="fas fa-cog fa-spin"></i>
        </li>
        
    </ul>
</div>

    </div>

    <div class="zoom-in-image-mask">
    <img class="zoom-in-image">
</div>


    
        <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
          <span class="search-input-field-pre">
            <i class="fas fa-keyboard"></i>
          </span>
            <div class="search-input-container">
                <input autocomplete="off"
                       autocorrect="off"
                       autocapitalize="off"
                       placeholder="搜索..."
                       spellcheck="false"
                       type="search"
                       class="search-input"
                >
            </div>
            <span class="close-popup-btn">
                <i class="fas fa-times"></i>
            </span>
        </div>
        <div id="search-result">
            <div id="no-result">
                <i class="fas fa-spinner fa-pulse fa-5x fa-fw"></i>
            </div>
        </div>
    </div>
</div>

    

</main>




<script src="/js/utils.js"></script>

<script src="/js/main.js"></script>

<script src="/js/header-shrink.js"></script>

<script src="/js/back2top.js"></script>

<script src="/js/dark-light-toggle.js"></script>





    
<script src="/js/local-search.js"></script>




    
<script src="/js/code-block.js"></script>




    
<script src="/js/lazyload.js"></script>



<div class="post-scripts pjax">
    
        
<script src="/js/post-helper.js"></script>

        
            
<script src="/js/libs/anime.min.js"></script>

        
        
            
<script src="/js/toc.js"></script>

        
    
</div>


    
<script src="/js/libs/pjax.min.js"></script>

<script>
    window.addEventListener('DOMContentLoaded', () => {
        window.pjax = new Pjax({
            selectors: [
                'head title',
                '.page-container',
                '.pjax'
            ],
            history: true,
            debug: false,
            cacheBust: false,
            timeout: 0,
            analytics: false,
            currentUrlFullReload: false,
            scrollRestoration: false,
            // scrollTo: true,
        });

        document.addEventListener('pjax:send', () => {
            KEEP.utils.pjaxProgressBarStart();
        });

        document.addEventListener('pjax:complete', () => {
            KEEP.utils.pjaxProgressBarEnd();
            window.pjax.executeScripts(document.querySelectorAll('script[data-pjax], .pjax script'));
            KEEP.refresh();
        });
    });
</script>



</body>
</html>
